receiver/opencensusreceiver/factory.go

```diff
@@ -0,0 +1,116 @@
+// limitations under the License.
+
+package opencensusreceiver
+
+import (
+	"context"
+
+	"github.com/census-instrumentation/opencensus-service/internal/configmodels"
+	"github.com/census-instrumentation/opencensus-service/internal/factories"
+
+	"github.com/census-instrumentation/opencensus-service/consumer"
+	"github.com/census-instrumentation/opencensus-service/receiver"
+)
+
+var _ = factories.RegisterReceiverFactory(&ReceiverFactory{})
+
+const (
+	// The value of "type" key in configuration.
+	typeStr = "opencensus"
+)
+
+// ReceiverFactory is the factory for receiver.
+type ReceiverFactory struct {
+}
+
+// Type gets the type of the Receiver config created by this factory.
+func (f *ReceiverFactory) Type() string {
+	return typeStr
+}
+
+// CustomUnmarshaler returns nil because we don't need custom unmarshaling for this config.
+func (f *ReceiverFactory) CustomUnmarshaler() factories.CustomUnmarshaler {
+	return nil
+}
+
+// CreateDefaultConfig creates the default configuration for receiver.
+func (f *ReceiverFactory) CreateDefaultConfig() configmodels.Receiver {
+	return &ConfigV2{
+		ReceiverSettings: configmodels.ReceiverSettings{
+			Endpoint: "127.0.0.1:55678",
+			Enabled:  true,
+		},
+	}
+}
+
+// CreateTraceReceiver creates a  trace receiver based on provided config.
+func (f *ReceiverFactory) CreateTraceReceiver(
+	ctx context.Context,
+	cfg configmodels.Receiver,
+	nextConsumer consumer.TraceConsumer,
+) (receiver.TraceReceiver, error) {
+
+	r, err := f.createReceiver(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	r.traceConsumer = nextConsumer
+
+	return r, nil
+}
+
+// CreateMetricsReceiver creates a metrics receiver based on provided config.
+func (f *ReceiverFactory) CreateMetricsReceiver(
+	cfg configmodels.Receiver,
+	consumer consumer.MetricsConsumer,
+) (receiver.MetricsReceiver, error) {
+
+	r, err := f.createReceiver(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	r.metricsConsumer = consumer
+
+	return r, nil
+}
+
+func (f *ReceiverFactory) createReceiver(cfg configmodels.Receiver) (*Receiver, error) {
+	rCfg := cfg.(*ConfigV2)
+
+	// There must be one receiver for both metrics and traces. We maintain a map of
+	// receivers per config.
+
+	// Check to see if there is already a receiver for this config.
+	receiver, ok := receivers[rCfg]
+	if !ok {
+		// We don't have a receiver, so create one.
+		var err error
+		receiver, err = New(rCfg.Endpoint, nil, nil)
+		if err != nil {
+			return nil, err
+		}
+		// Remember the receiver in the map
+		receivers[rCfg] = receiver
+	}
+	return receiver, nil
+}
+
+// This is the map of already created OpenCensus receivers for particular configurations.
+// We maintain this map because the factory is asked trace and metric receivers separately
+// when it gets CreateTraceReceiver() and CreateMetricsReceiver() but they must not
+// create separate objects, they must use one Receiver object per configuration.
+var receivers = map[*ConfigV2]*Receiver{}
```

Comment:
q.: so the caller code of Create needs to ensure that the same config object is used, how that is going to be enforced?

Reply:
It doesn't need to be enforced. The idea is that _if_ the caller (the pipeline builder) attempts to create a receiver using the same config instance the factory will end up returning the same receiver instance which is the desirable effect.